File: 0131-palindrome-partitioning/0131-palindrome-partitioning.java
class Solution {
    List<List<String>> result;
    public List<List<String>> partition(String s) {
        result = new ArrayList<>();

        helper(s, 0, new ArrayList<>());

        return result;
    }

    private void helper(String s, int index, List<String> path){
        // System.out.println(path);
        if(index == s.length()){
            result.add(new ArrayList<>(path));
            return;
        }

        for(int i = index; i < s.length(); i++){
            if(isPalindrome(s, index, i) == true){
                // StringBuilder temp = new StringBuilder(s.substring(index, i));
                path.add(s.substring(index, i+1));
                helper(s, i + 1, path);
                path.remove(path.size() - 1);

            }
        }
    }

    private boolean isPalindrome(String s, int left, int right){

        while(left < right){
            if(s.charAt(left) != s.charAt(right)){
                return false;
            }
            left++;
            right--;
        }

        return true;
    }
}

//For Loop backtrack Recursion 

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.py
class Solution:
    result = []
    def partition(self, s: str) -> List[List[str]]:
        self.result = list()
        if s is None or len(s) == 0:
            return self.result

        self.helper(s, list())
        return self.result
    
    def helper(self, s, path):
        if len(s) == 0:
            self.result.append(path.copy())
        
        for i in range(0, len(s)):
            if self.isPalindrome(s, 0, i):
                path.append(s[0 : i + 1])
                self.helper(s[i+1:], path)
                path.pop(-1)

    def isPalindrome(self, s, start, end):

        while start < end:
            if s[start] != s[end]:
                return False
            
            start += 1
            end -= 1
        
        return True

# Substring For loop backtrack recursion
# Time Complexity: O(n * 2^n)
# Space Complexity: O()